required/latex-lab/latex-lab-math.dtx

```diff
@@ -1463,6 +1473,164 @@
 %    \end{macrocode}
 % \end{macro}
 %
+% \subsection{Token-by-token inline grabbing}
+%
+% Grabbing inline math token-by-token is more involved. The mechanism here
+% is essentially a simplified version of that originally seen in
+% \pkg{collcell} and refined in \pkg{siunitx}. We make use of the fact that
+% in math mode spaces are ignored, so we have to deal with only \texttt{N}-type
+% tokens and groups. Furthermore, there is no need to look inside groups, so
+% the only special cases are a small selection of \texttt{N}-type tokens.
+%
+% \begin{variable}{\l_@@_grabbed_tl}
+%   For collection of the material piecewise.
+%    \begin{macrocode}
+\tl_new:N \l_@@_grabbed_tl
+%    \end{macrocode}
+% \end{variable}
+%
+% \begin{variable}{\l_@@_grab_env_int}
+%   Needed to count up the number of nested environments encountered.
+%    \begin{macrocode}
+\int_new:N \l_@@_grab_env_int
+%    \end{macrocode}
+% \end{variable}
+%
+% \begin{macro}{\@@_grab_dollar_loop:}
+% \begin{macro}{\@@_grab_loop:}
+%   The lead-off here establishes a group: we need that as we will have to
+%   be careful in the way \tn{cr} is handled and ensure this is only
+%   manipulated whilst grabbing. The main loop is then started.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_dollar_loop:
+  {
+    \group_begin:
+      \tl_clear:N \l_@@_grabbed_tl
+      \@@_grab_loop:
+  }
+\cs_new_protected:Npn \@@_grab_loop:
+  {
+    \peek_remove_spaces:n
+      {
+        \peek_meaning:NTF \c_group_begin_token
+          { \@@_grab_loop_group:n }
+          { \@@_grab_loop_token:N }
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+%
+% \begin{macro}{\@@_grab_loop_group:n}
+% \begin{macro}{\@@_grab_loop_store:n}
+%   Handling of grabbed groups is pretty easy.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_loop_group:n #1
+  { \@@_grab_loop_store:n { {#1} } }
+\cs_new_protected:Npn \@@_grab_loop_store:n #1
+  {
+    \tl_put_right:Nn \l_@@_grabbed_tl {#1}
+    \@@_grab_loop:
+  }
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+%
+% \begin{macro}{\@@_grab_loop_token:N}
+% \begin{macro}
+%   {
+%      \@@_grab_loop_$:              ,
+%      \@@_grab_loop_\\:             ,
+%      \@@_grab_loop_\begin:         ,
+%      \@@_grab_loop_\begin :        ,
```

Comment:
thatlooks pretty odd ... warrants an explanation? or is it a mistake?

Reply:
If `\begin` is at the start of a cell, it expands to `\protect\begin[space]` - the reason I asked if we want to 'tidy up'. Unless we do, there has to be coverage of the space-added macro

Reply:
not sure how you will tidy that up, which is why I suggested explaining the case

Reply:
Tidy up is as I said earlier: find a `\protect` token, grab the next token, see if it is `\foo[space]` and `\foo` exists, process just `\foo`.

Reply:
or make `\begin` etex protected rather than `\protect` ....

Reply:
@davidcarlisle Oh yes _please_

Reply:
maybe that indeed.

Reply:
I've explained in a comment why the case is covered at present: if we can switch to `\protected\def` for `\begin`, this special case could be removed.